### PR TITLE
Demand objective functions

### LINKED ISSUE #168: Create objective functions relating the electricity demand to the temperature for each month
The following figure shows the objective function (linear regression) that correlates the electricity demand to the wet-bulb temperature.

demandtempcorr

This issue can be closed when such analysis, and also the same analysis for each month is added to the repo.

### PR DESCRIPTION
**Pull request by robfairh**
Fixes #168

### CODE REVIEW

File: data_processing/grid_demand_objective_function.ipynb [hunk 1]
```diff
@@ -0,0 +1,817 @@
...
+    "                 skiprows=1,\n",
+    "                 parse_dates=True,\n",
+    "                 usecols=['Date and Time', 'Value (KW)'],\n",
+    "                 index_col = ['Date and Time'])\n",
+    "demd2019 = df.rename(columns={\"Value (KW)\":\"kW\"})\n",
+    "\n",
+    "df = pd.read_csv(\"willard_weather_data.csv\",\n",
+    "                 parse_dates=True, usecols=['DATE', 'HourlyWetBulbTemperature'],\n",
+    "                 index_col = ['DATE'])\n",
+    "df = df.rename(columns={\"HourlyWetBulbTemperature\":\"temperature\"})  # temperature in F\n",
```
Comment (yardasol) [inline data_processing/grid_demand_objective_function.ipynb:160]: I think it's important to keep the fact that this temperature is a wet bulb temperature.

Comment (yardasol) [reply data_processing/grid_demand_objective_function.ipynb:160]:
```suggestion
    "df = df.rename(columns={\"HourlyWetBulbTemperature\":\"wb_temperature\"})  # wet bulb temperature in F\n",
```

File: data_processing/grid_demand_objective_function.ipynb [hunk 2]
```diff
@@ -0,0 +1,817 @@
...
+    "temp2019 = df['06/30/2018 23:53':'06/30/2019 23:53'].dropna()\n",
+    "\n",
+    "temp_daterange = pd.date_range(start='06/30/2018 23:53', end='06/30/2019 22:53', freq='h')\n",
+    "demd_daterange = demd2019.index\n",
+    "\n",
+    "# remove values outside temp_daterange and flag empty values\n",
+    "temp = []\n",
+    "for date in temp_daterange:\n",
+    "    if (temp2019.index == date).any() == True:\n",
+    "        addtempt = temp2019[temp2019.index == date].to_string(columns=['temperature'], header=False, index=False)\n",
```
Comment (yardasol) [inline data_processing/grid_demand_objective_function.ipynb:170]:
```suggestion
    "        addtempt = temp2019[temp2019.index == date].to_string(columns=['wb_temperature'], header=False, index=False)\n",
```

File: data_processing/grid_demand_objective_function.ipynb [hunk 3]
```diff
@@ -0,0 +1,817 @@
...
+    "        temp.append(float(addtempt))\n",
+    "    else:\n",
+    "        temp.append('empty')\n",
+    "\n",
+    "temp2019 = pd.DataFrame(temp, index=demd_daterange, columns=['temperature'])\n",
```
Comment (yardasol) [inline data_processing/grid_demand_objective_function.ipynb:175]:
```suggestion
    "temp2019 = pd.DataFrame(temp, index=demd_daterange, columns=['wb_temperature'])\n",
```

File: data_processing/grid_demand_objective_function.ipynb [hunk 4]
```diff
@@ -0,0 +1,817 @@
...
+      ]
+     },
+     "metadata": {
+      "needs_background": "light"
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "# plot demand along with the temperature\n",
```
Comment (yardasol) [inline data_processing/grid_demand_objective_function.ipynb]:
```suggestion
    "# plot demand along with the wet bulb temperature\n",
```

File: data_processing/grid_demand_objective_function.ipynb [hunk 5]
```diff
@@ -0,0 +1,817 @@
...
+     },
+     "execution_count": 3,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# merge dataframes and remove empty values\n",
+    "df = pd.concat([demd2019, temp2019], axis=1)\n",
+    "df = df[df.temperature != 'empty']\n",
```
Comment (yardasol) [inline data_processing/grid_demand_objective_function.ipynb:301]:
```suggestion
    "df = df[df.wb_temperature != 'empty']\n",
```

File: data_processing/grid_demand_objective_function.ipynb [hunk 6]
```diff
@@ -0,0 +1,817 @@
...
+      "needs_background": "light"
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "# plot demand along with the temperature\n",
+    "fig, ax = plt.subplots(figsize=(14,8))\n",
+    "df.kW.plot(ax=ax)\n",
+    "df.temperature.plot(ax=ax, secondary_y=True)\n",
```
Comment (yardasol) [inline data_processing/grid_demand_objective_function.ipynb:337]:
```suggestion
    "df.wb_temperature.plot(ax=ax, secondary_y=True)\n",
```

File: data_processing/grid_demand_objective_function.ipynb [hunk 7]
```diff
@@ -0,0 +1,817 @@
...
+    "ax.legend([ax.get_lines()[0], ax.right_ax.get_lines()[0]], ['demand','temperature'], bbox_to_anchor=(1.17, 1.0))"
```
Comment (yardasol) [inline data_processing/grid_demand_objective_function.ipynb]:
```suggestion
    "ax.legend([ax.get_lines()[0], ax.right_ax.get_lines()[0]], ['demand','wet bulb temperature'], bbox_to_anchor=(1.17, 1.0))"
```

File: data_processing/grid_demand_objective_function.ipynb [hunk 8]
```diff
@@ -0,0 +1,817 @@
...
+     },
+     "metadata": {
+      "needs_background": "light"
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "ax = df.plot.scatter(x='temperature', y='kW', figsize=(14, 8))\n",
+    "ax.set_xlabel(\"Temperature [F]\")\n",
```
Comment (yardasol) [inline data_processing/grid_demand_objective_function.ipynb]:
```suggestion
    "ax.set_xlabel(\"Wet Bulb Temperature [F]\")\n",
```

File: data_processing/grid_demand_objective_function.ipynb [hunk 9]
```diff
@@ -0,0 +1,817 @@
...
+      ]
+     },
+     "metadata": {
+      "needs_background": "light"
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "ax = df.plot.scatter(x='temperature', y='kW', figsize=(14, 8))\n",
```
Comment (yardasol) [inline data_processing/grid_demand_objective_function.ipynb:372]:
```suggestion
    "ax = df.plot.scatter(x='wb_temperature', y='kW', figsize=(14, 8))\n",
```

File: data_processing/grid_demand_objective_function.ipynb [hunk 10]
```diff
@@ -0,0 +1,817 @@
...
+    "    None\n",
+    "    '''\n",
+    "\n",
+    "    train_X = dataframe['temperature'].to_numpy().reshape(-1, 1)\n",
+    "    train_Y = dataframe['kW'].to_numpy().reshape(-1, 1)\n",
+    "\n",
+    "    # plot original data\n",
+    "    plt.figure(figsize=(14, 8))\n",
+    "    plt.scatter(train_X, train_Y, label='Data')\n",
+    "    plt.xlabel(\"Temperature [F]\")\n",
```
Comment (yardasol) [inline data_processing/grid_demand_objective_function.ipynb]:
```suggestion
    "    plt.xlabel(\"Wet Bulb Temperature [F]\")\n",
```

File: data_processing/grid_demand_objective_function.ipynb [hunk 11]
```diff
@@ -0,0 +1,817 @@
...
+    "    order: [int]\n",
+    "        polynomial regression order\n",
+    "    title: [string]\n",
+    "        title of the plot\n",
+    "    Returns:\n",
+    "    --------\n",
+    "    None\n",
+    "    '''\n",
+    "\n",
+    "    train_X = dataframe['temperature'].to_numpy().reshape(-1, 1)\n",
```
Comment (yardasol) [inline data_processing/grid_demand_objective_function.ipynb:408]:
```suggestion
    "    train_X = dataframe['wb_temperature'].to_numpy().reshape(-1, 1)\n",
```

File: data_processing/grid_demand_objective_function.ipynb [hunk 12]
```diff
@@ -0,0 +1,817 @@
...
+   "source": [
+    "def demand_vs_temperature_polynomial_regression(dataframe, order, title):\n",
+    "    '''\n",
+    "    Produces a scattered plot with the orignal data and a regression line.\n",
+    "\n",
+    "    Parameters:\n",
+    "    -----------\n",
+    "    dataframe: [pandas dataframe]\n",
+    "        index: date and time\n",
+    "        columns: kw, temperature\n",
```
Comment (yardasol) [inline data_processing/grid_demand_objective_function.ipynb:398]:
```suggestion
    "        columns: kw, wb_temperature\n",
```

File: duck-curve/duck-curve.ipynb [hunk 13]
```diff
@@ -612,7 +612,7 @@
    "name": "python",
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.7.6"
+   "version": "3.8.3"
```
Comment (yardasol) [inline duck-curve/duck-curve.ipynb:615]: yay for newer python!

**Review by yardasol — COMMENTED** [review]
Nice work @robfairh. The only thing that stood out to me was that you hid the fact the the temperature we are using in this dataset is a wet bulb temperature. I've added suggestions on where you this could be specified. I think that at the minimum the plot labels and titles should reference that the temperature is a wet bulb temperature.

- [x] Read the PR description
- [x] Read through all the changes, considering the following questions.
  - [x] Is there anything you can praise about this PR? Start with
        praise.
  - [x] Are variable names brief but descriptive?
  - [ ] Are new/changed functions no longer than a paragraph?
  - [ ] Do all function parameters have default values where appropriate?
  - [x] Is the code clear and clean? (see Robert C. Martin's
        [Clean Code](https://i-share-uiu.primo.exlibrisgroup.com/discovery/fulldisplay?docid=alma99944155312205899&context=L&vid=01CARLI_UIU:CARLI_UIU&tab=LibraryCatalog&lang=en))
  - [x] Is there enough documentation?
  - [ ] Does the programming style meet the requirements of the
        repository ([PEP8](https://www.python.org/dev/peps/pep-0008/) for python,
        [google C++ style guide](https://google.github.io/styleguide/cppguide.html), etc.)
  - [ ] If a new feature has been added, or a bug fixed, has a test been
        added to confirm good behavior?
  - [ ] Does the test actually test the new/changed functionality?
  - [ ] Does the test successfully test edge cases?
  - [ ] Does the test successfully test corner cases?
  - [ ] If the repository has continuous integration: does the PR pass
        the tests?
  - [x] If there is no continuous integration, check out the branch
        locally, build, and run the tests.
  - [ ] Do the tests pass on your machine?
  - [x] Is the PR free of random cruft (built files, `.swp` files,
        etc.)?
  - [x] Do all files in the PR belong in the repository?
  - [ ] If the PR deletes files, is this appropriate?
  - [x] If the PR adds files or data, are these new files compatible
        with the repository license?
  - [x] Make a review, leaving kind comments and suggesting changes
        where needed (to resolve the above).
  - [x] Has the author resolved all of the comments and suggestions
        in your review?
- [x] When you approve of the PR, merge and close it.
- [x] Does this PR close an issue? If so, be sure to descriptively
      close this issue when the PR is merged
- [x] Thank the author for their contribution.

File: data_processing/grid_demand_objective_function.ipynb (hunk 1 above)
Comment (robfairh) [reply data_processing/grid_demand_objective_function.ipynb:160]: I want to leave this generic in case future modifications include another type of temperature.
However, I will accept the changes on the plots' labels/titles.

**Comment by robfairh** [comment]
Thanks for your review @yardasol ! I have accepted some of your suggestions.
The PR is ready to be re-reviewed/merged accordingly.

**Review by yardasol — APPROVED** [review]
Excelente!
